libraries/AP_InertialSensor/AP_InertialSensor_SCH16T.cpp

```diff
@@ -0,0 +1,472 @@
+                wait_us = 5000UL;
+                break;
+            }
+
+        case State::Validate: {
+                ReadStatusRegisters(); // Read all status registers twice
+                ReadStatusRegisters();
+
+                // Check that registers are configured properly and that the sensor status is OK
+                if (ValidateSensorStatus() && ValidateRegisterConfiguration()) {
+                    hal.console->printf("Starting read loop");
+                    _state = State::Read;
+                    wait_us = 1000000UL / expected_sample_rate_hz;
+                    dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+
+                } else {
+                    hal.console->printf("Sensor or Register validation failed, resetting");
+                    _state = State::Reset;
+                    wait_us = 100000UL;
+                }
+
+                break;
+            }
+
+        case State::Read: {
+                uint32_t tstart = AP_HAL::micros();
+
+                if (drdy_pin != 0) {
+                    wait_ok = hal.gpio->wait_pin(drdy_pin, AP_HAL::GPIO::INTERRUPT_RISING, 2100);
+                }
+
+                // Collect the data
+                SensorData data = {};
+
+                if (wait_ok && ReadData(&data)) {
+
+                    Vector3f accel{accel_scale*data.acc_x, accel_scale*data.acc_y, accel_scale*data.acc_z};
+                    Vector3f gyro{gyro_scale*data.gyro_x, gyro_scale*data.gyro_y, gyro_scale*data.gyro_z};
+
+                    _rotate_and_correct_accel(accel_instance, accel);
+                    _notify_new_accel_raw_sample(accel_instance, accel);
+
+                    _rotate_and_correct_gyro(gyro_instance, gyro);
+                    _notify_new_gyro_raw_sample(gyro_instance, gyro);
+
+                    _publish_temperature(accel_instance, float(data.temp)/100.f);
+
+                    // Calculate next reschedule
+                    const uint32_t period_us = (1000000UL / expected_sample_rate_hz) - 20U;
+                    uint32_t dt = AP_HAL::micros() - tstart;
+                    if (dt < period_us) {
+                        wait_us = period_us - dt;
+                    }
+
+                    if (failure_count > 0) {
+                        failure_count--;
+                    }
+
+                } else {
+                    failure_count++;
+                }
+
+                // Reset if successive failures
+                if (failure_count > 10) {
+                    hal.console->printf("Failure count high, resetting");
+                    _state = State::Reset;
+                    return;
+                }
+
+                break;
+            }
+
+        default:
+            break;
+        } // end switch/case
+
+        // reschedule
+        hal.scheduler->delay_microseconds(wait_us);
+    }
+}
+
+bool AP_InertialSensor_SCH16T::ReadData(SensorData *data)
+{
+    RegisterRead(RATE_X2);
+    uint64_t gyro_x = RegisterRead(RATE_Y2);
```

Comment:
Looks like these are consecutive. You can used a packed structure and read the whole frame in one go. This will be faster as well as saving flash

Reply:
the SafeSpi protocol doesn't work that way, all transfers are 48bit out-of-frame, out of frame meaning the response is contained in the rx of the following transfer